File: ait-node/src/agent.js
import { fileURLToPath } from "node:url";

import {
  AGENT_CAPABILITIES_CONTRACT,
  SUPPORTED_TRANSPORTS,
  SUPPORTED_WORKER_OPERATIONS,
} from "./contract.js";
import { NativeProtocolError } from "./errors.js";
import { NativeRuntime } from "./runtime.js";

const TRANSPORTS = new Set(SUPPORTED_TRANSPORTS);
const WORKER_OPERATIONS = new Set(SUPPORTED_WORKER_OPERATIONS);
const WORKER_CONTEXT_FIELDS = Object.freeze({
  cwd: "cwd",
  repoRoot: "repo_root",
  manifestPath: "manifest_path",
  env: "env",
  signature: "signature",
  signatureTimestamp: "signature_timestamp",
  nowUnixSeconds: "now_unix_seconds",
});

export class AgentCapabilities {
  constructor(payload) {
    if (!isRecord(payload)) {
      throw new NativeProtocolError(
        "ait-agent-worker capabilities must be a JSON object",
      );
    }
    if (payload.contract !== AGENT_CAPABILITIES_CONTRACT) {
      throw new NativeProtocolError(
        "ait-agent-worker returned an unsupported capabilities contract",
      );
    }
    if (payload.binary !== "ait-agent-worker") {
      throw new NativeProtocolError(
        "ait-agent-worker capabilities identify the wrong binary",
      );
    }
    if (payload.python_worker_execution_allowed !== false) {
      throw new NativeProtocolError(
        "ait-agent-worker capabilities permit a forbidden Python fallback",
      );
    }

    const supportedTransports = requiredTextArray(
      payload,
      "supported_transports",
    );
    const unknown = supportedTransports.filter((value) => !TRANSPORTS.has(value));
    if (unknown.length > 0) {
      throw new NativeProtocolError(
        `ait-agent-worker reported unsupported transports: ${unknown
          .sort()
          .join(", ")}`,
      );
    }
    const eventLoopBackends = requiredTextArray(
      payload,
      "event_loop_backends",
    );
    const defaultEventLoopBackend = requiredText(
      payload,
      "default_event_loop_backend",
    );
    if (!eventLoopBackends.includes(defaultEventLoopBackend)) {
      throw new NativeProtocolError(
        "ait-agent-worker default event-loop backend is not available",
      );
    }

    this.contract = AGENT_CAPABILITIES_CONTRACT;
    this.version = requiredText(payload, "version");
    this.platform = requiredText(payload, "platform");
    this.architecture = requiredText(payload, "architecture");
    this.supportedTransports = Object.freeze(supportedTransports);
    this.eventLoopBackends = Object.freeze(eventLoopBackends);
    this.defaultEventLoopBackend = defaultEventLoopBackend;
    this.raw = payload;
    Object.freeze(this);
  }
}

export class AgentClient {
  constructor(runtime = new NativeRuntime()) {
    if (!(runtime instanceof NativeRuntime)) {
      throw new TypeError("runtime must be a NativeRuntime");
    }
    this.runtime = runtime;
  }

  capabilities() {
    return new AgentCapabilities(this.runtime.agentCapabilities());
  }

  workerTransaction(operation, payload, options = {}) {
    if (typeof operation !== "string") {
      throw new TypeError("worker operation must be text");
    }
    const normalizedOperation = operation.trim();
    if (!WORKER_OPERATIONS.has(normalizedOperation)) {
      throw new TypeError(
        `unsupported worker operation ${JSON.stringify(operation)}; expected: ${[
          ...WORKER_OPERATIONS,
        ]
          .sort()
          .join(", ")}`,
      );
    }
    if (payload === undefined) {
      throw new TypeError("worker payload is required");
    }
    const normalizedOptions = optionsObject(options, "worker options");
    const { worker = "main", ...context } = normalizedOptions;
    return this.runtime.agentWorkerTransaction({
      operation: normalizedOperation,
      payload,
      worker: normalizeWorkerName(worker),
      ...workerContext(context),
    });
  }

  slackCommand(payload, options = {}) {
    return this.workerTransaction("slack-command", payload, options);
  }

  discordInteraction(payload, options = {}) {
    return this.workerTransaction("discord-interaction", payload, options);
  }

  replyProvider(payload, options = {}) {
    return this.workerTransaction("reply-provider", payload, options);
  }

}

function workerContext(options) {
  return bindingContext(options, WORKER_CONTEXT_FIELDS, "worker options");
}

function bindingContext(options, fields, label) {
  const values = optionsObject(options, label);
  const unknown = Object.keys(values).filter(
    (name) => !Object.hasOwn(fields, name),
  );
  if (unknown.length > 0) {
    throw new TypeError(
      `unsupported ${label} fields: ${unknown.sort().join(", ")}`,
    );
  }
  const result = {};
  for (const [publicName, requestName] of Object.entries(fields)) {
    const value = values[publicName];
    if (value === undefined || value === null) {
      continue;
    }
    if (publicName === "env") {
      result[requestName] = normalizeEnvironment(value);
    } else if (publicName === "nowUnixSeconds") {
      if (!Number.isSafeInteger(value)) {
        throw new TypeError("nowUnixSeconds must be a safe integer");
      }
      result[requestName] = value;
    } else {
      result[requestName] = normalizeTextOrFileUrl(value, publicName);
    }
  }
  return result;
}

function normalizeEnvironment(value) {
  if (!isRecord(value)) {
    throw new TypeError("env must be an object");
  }
  const result = {};
  for (const [name, item] of Object.entries(value)) {
    if (name.length === 0) {
      throw new TypeError("environment override names must be non-empty");
    }
    if (item === null || item === undefined) {
      result[name] = null;
    } else if (
      typeof item === "string" ||
      typeof item === "number" ||
      typeof item === "boolean"
    ) {
      result[name] = String(item);
    } else {
      throw new TypeError(
        `environment override ${name} must be text, number, boolean, or null`,
      );
    }
  }
  return result;
}

function normalizeTextOrFileUrl(value, field) {
  const normalized =
    value instanceof URL ? fileURLToPath(value) : String(value).trim();
  if (normalized.length === 0) {
    throw new TypeError(`${field} must not be empty`);
  }
  return normalized;
}

function normalizeWorkerName(value) {
  if (typeof value !== "string") {
    throw new TypeError("worker name must be text");
  }
  const normalized = value.trim();
  if (normalized.length === 0) {
    throw new TypeError("worker name must not be empty");
  }
  return normalized;
}

function requiredText(payload, field) {
  const value = payload[field];
  if (typeof value !== "string" || value.length === 0) {
    throw new NativeProtocolError(
      `ait-agent-worker capabilities field ${field} must be non-empty text`,
    );
  }
  return value;
}

function requiredTextArray(payload, field) {
  const value = payload[field];
  if (
    !Array.isArray(value) ||
    value.length === 0 ||
    value.some((item) => typeof item !== "string" || item.length === 0) ||
    new Set(value).size !== value.length
  ) {
    throw new NativeProtocolError(
      `ait-agent-worker capabilities field ${field} must be unique text`,
    );
  }
  return [...value];
}

function optionsObject(value, label) {
  if (!isRecord(value)) {
    throw new TypeError(`${label} must be an object`);
  }
  return value;
}

function isRecord(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

File: ait-python/src/ait_python/agent.py
from __future__ import annotations

from dataclasses import dataclass
import os
from typing import Any, Mapping

from .contract import AGENT_CAPABILITIES_CONTRACT, SUPPORTED_TRANSPORTS
from .errors import NativeProtocolError
from .runtime import NativeRuntime


JsonObject = Mapping[str, Any]
_WORKER_OPERATIONS = frozenset(
    {"slack-command", "discord-interaction", "reply-provider"}
)


@dataclass(frozen=True, slots=True)
class AgentCapabilities:
    contract: str
    version: str
    platform: str
    architecture: str
    supported_transports: tuple[str, ...]
    event_loop_backends: tuple[str, ...]
    default_event_loop_backend: str
    raw: JsonObject

    @classmethod
    def from_payload(cls, payload: Any) -> AgentCapabilities:
        if not isinstance(payload, dict):
            raise NativeProtocolError(
                "ait-agent-worker capabilities must be a JSON object"
            )
        if payload.get("contract") != AGENT_CAPABILITIES_CONTRACT:
            raise NativeProtocolError(
                "ait-agent-worker returned an unsupported capabilities contract"
            )
        if payload.get("binary") != "ait-agent-worker":
            raise NativeProtocolError(
                "ait-agent-worker capabilities identify the wrong binary"
            )
        if payload.get("python_worker_execution_allowed") is not False:
            raise NativeProtocolError(
                "ait-agent-worker capabilities permit a forbidden Python fallback"
            )

        transports = _required_text_list(payload, "supported_transports")
        unknown = set(transports).difference(SUPPORTED_TRANSPORTS)
        if unknown:
            names = ", ".join(sorted(unknown))
            raise NativeProtocolError(
                f"ait-agent-worker reported unsupported transports: {names}"
            )
        backends = _required_text_list(payload, "event_loop_backends")
        default_backend = _required_text(payload, "default_event_loop_backend")
        if default_backend not in backends:
            raise NativeProtocolError(
                "ait-agent-worker default event-loop backend is not available"
            )
        return cls(
            contract=AGENT_CAPABILITIES_CONTRACT,
            version=_required_text(payload, "version"),
            platform=_required_text(payload, "platform"),
            architecture=_required_text(payload, "architecture"),
            supported_transports=transports,
            event_loop_backends=backends,
            default_event_loop_backend=default_backend,
            raw=payload,
        )


@dataclass(frozen=True, slots=True)
class AgentClient:
    runtime: NativeRuntime = NativeRuntime()

    def capabilities(self) -> AgentCapabilities:
        return AgentCapabilities.from_payload(self.runtime.agent_capabilities())

    def worker_transaction(
        self,
        operation: str,
        payload: Any,
        *,
        worker: str = "main",
        **context: Any,
    ) -> Any:
        normalized = operation.strip()
        if normalized not in _WORKER_OPERATIONS:
            choices = ", ".join(sorted(_WORKER_OPERATIONS))
            raise ValueError(
                f"unsupported worker operation {operation!r}; expected: {choices}"
            )
        return self.runtime.agent_worker_transaction(
            {
                "operation": normalized,
                "payload": payload,
                "worker": _worker_name(worker),
                **_context(context),
            }
        )

    def slack_command(
        self, payload: Any, *, worker: str = "main", **context: Any
    ) -> Any:
        return self.worker_transaction(
            "slack-command", payload, worker=worker, **context
        )

    def discord_interaction(
        self, payload: Any, *, worker: str = "main", **context: Any
    ) -> Any:
        return self.worker_transaction(
            "discord-interaction", payload, worker=worker, **context
        )

    def reply_provider(
        self, payload: Any, *, worker: str = "main", **context: Any
    ) -> Any:
        return self.worker_transaction(
            "reply-provider", payload, worker=worker, **context
        )

def _context(values: Mapping[str, Any]) -> dict[str, Any]:
    aliases = {
        "cwd": "cwd",
        "repo_root": "repo_root",
        "manifest_path": "manifest_path",
        "env": "env",
        "signature": "signature",
        "signature_timestamp": "signature_timestamp",
        "now_unix_seconds": "now_unix_seconds",
    }
    unknown = set(values).difference(aliases)
    if unknown:
        names = ", ".join(sorted(unknown))
        raise TypeError(f"unsupported agent context fields: {names}")
    result: dict[str, Any] = {}
    for public_name, request_name in aliases.items():
        if public_name not in values or values[public_name] is None:
            continue
        value = values[public_name]
        if public_name == "env":
            result[request_name] = _environment(value)
        elif public_name == "now_unix_seconds":
            if not isinstance(value, int) or isinstance(value, bool):
                raise TypeError("now_unix_seconds must be an integer")
            result[request_name] = value
        else:
            result[request_name] = os.fspath(value)
    return result


def _environment(value: Any) -> dict[str, str | None]:
    if not isinstance(value, Mapping):
        raise TypeError("env must be a mapping")
    result: dict[str, str | None] = {}
    for name, item in value.items():
        if not isinstance(name, str) or not name:
            raise TypeError("environment override names must be non-empty strings")
        result[name] = None if item is None else os.fspath(item)
    return result


def _worker_name(value: str) -> str:
    normalized = value.strip()
    if not normalized:
        raise ValueError("worker name must not be empty")
    return normalized


def _required_text(payload: JsonObject, field: str) -> str:
    value = payload.get(field)
    if not isinstance(value, str) or not value:
        raise NativeProtocolError(
            f"ait-agent-worker capabilities field {field} must be non-empty text"
        )
    return value


def _required_text_list(payload: JsonObject, field: str) -> tuple[str, ...]:
    value = payload.get(field)
    if (
        not isinstance(value, list)
        or not value
        or any(not isinstance(item, str) or not item for item in value)
        or len(set(value)) != len(value)
    ):
        raise NativeProtocolError(
            f"ait-agent-worker capabilities field {field} must be unique text"
        )
    return tuple(value)
